### PR TITLE
[WIP] Add documentation for heap snapshot schema

### PR DESCRIPTION
**Pull request by joselea**

### CODE REVIEW

File: microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+The DevTools Memory tool allows you to explore all the objects allocated in memory by taking a heap snapshot. This information is useful for performance investigations because you can find out what objects are consuming the most memory. Read more on [Heap Snapshots](heap-snapshots.md). 
+
+In DevTools the results are shown in the form of a table, in the top there is a list of aggerates that summarizes all the objects by type or class name and in the bottom there is a list of retainers which gets updated every time you elect a single object from the list of aggregates. 
+
+Memory in v8 is a graph made of nodes and edges. The purpose of the `*.heapsnapshot` file is to represent that graph efficiently and to facilitate its transfer in chunks between the browser process and DevTools, so the contents of this file are a flattened representation of those node-edge relations as simple arrays of numbers and strings. The file has a `*.heapsnapshot` extension but its contents are actually just a `JSON`.
```
Comment (captainbrosset) [inline microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md:18]: This is probably for fairly advanced users anyway, but I wonder if we should give a short explanation about graphs, nodes and edges? Maybe with a little drawing too. Maybe a link to a wikipedia page like https://en.wikipedia.org/wiki/Graph_(abstract_data_type) .

File: microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+Array | Various internal lists, which show up in the devtools with the category name "(array)". Like Hidden, this category clumps together a bunch of unrelated things. Many of the objects here are named "(object properties)" or "(object elements)", indicating that they contain the string-keyed or numeric-keyed properties of a JavaScript object.
+Code | Things that grow proportionally to the amount of script, and/or the number of times that functions run..  
+Synthetic | Synthetic nodes don't correspond to anything actually allocated in memory. As far as I can tell, these are used only to distinguish the different kinds of GC roots.
+
+
+## Edges
+
+Similar to the **Nodes** array, edges contains all the individual elements needed to re-construct the edges and we can calculate the number of elements of this array by multiplying `edge_count` * `metadata.edge_fields.length`.
+
+Edges are stored sequentially and to read them properly you need to parse the nodes first, because each node knows how many edges it has.
```
Comment (captainbrosset) [inline microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md:182]: Reading this sentence makes me think that we might need a screenshot somewhere that explains how the data is organized in the file. This would be useful for visual learners. This sentence is quite hard to parse and a picture can be worth a thousand words.
Do you think there's an easy way to "draw" a little heapsnapshot sample to make things clearer?

File: microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+Synthetic | Synthetic nodes don't correspond to anything actually allocated in memory. As far as I can tell, these are used only to distinguish the different kinds of GC roots.
+
+
+## Edges
+
+Similar to the **Nodes** array, edges contains all the individual elements needed to re-construct the edges and we can calculate the number of elements of this array by multiplying `edge_count` * `metadata.edge_fields.length`.
+
+Edges are stored sequentially and to read them properly you need to parse the nodes first, because each node knows how many edges it has. 
+
+To re-construct an edge we need 3 elements of information, and we can use `metadata.edge_types` to understand what each element is.
```
Comment (captainbrosset) [inline microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md:184]: Please give the names of the 3 elements that are needed right after "we need 3 elements of information". Like "we need 3 elements of information: the edge type, its name or index, and the node it's connected to".

File: microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+type | See `edge_types[0]` to find out what are the possible types
+name_or_index | This could be a number or a string. If it’s a number, it indicates the index of the strings array where the name is stored.
+to_node | Index in the array of the node that it points to
+
+
+`edge_types` | Description  
+:- | :- 
+Internal | Edges that don't correspond to JS-visible names but are still important. As an example, Function instances have a "context" representing the state of variables that were in scope where the function was defined. There is no way for JS code to directly read the "context" of a function, but it's still a very important link when trying to understand retainers.
+Weak | Weak edges don't keep the target alive, and thus are omitted from the Retainers view. Any object with only weak edges pointing to it can be discarded by the GC.
+Hidden | Just like Internal, except these edges don't have nice names because nobody bothered to write special-case code for them.  They're just numbered in increasing order.
```
Comment (captainbrosset) [inline microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md:200]: The "nobody bothered to write..." part feels a bit harsh to me :) 
Maybe something like "Like Internal, except these edges don't have unique names and are just numbered in increasing order".

File: microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+Shortcut | An easier-to-read representation of some other path. This type is used very little. An example is if you use Function.prototype.bind to create a bound function with some bound arguments, V8 creates a JSBoundFunction, which points to a FixedArray (an internal type), which points to each bound argument. When producing a snapshot, V8 adds a shortcut edge from the bound function directly to each bound argument, bypassing the FixedArray.
+Element | Object properties where the key is a number.
+
+
+<!-- ====================================================================== -->
+## Sample code
+Now that you understand the format of this file you can create your own custom tool to extract data, run an analysis or create a visualization of the memory graph.  You can find sample code under this GitHub repo to get you started. It creates a force directed graph visualization.
```
Comment (captainbrosset) [inline microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md:207]: The link to the github repo is missing here.

File: microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+description: Use the Detached Elements tool in the Microsoft Edge DevTools to find and fix DOM memory leaks.
+author: MSEdgeTeam
+ms.author: joselea
+ms.topic: conceptual
+ms.prod: microsoft-edge
+ms.date: 9/13/2022
+---
+# Heap Snapshot Schema
+
+_In this document we will explain the schema of a `*.heapsnapshot` file which you can take and export using DevTools. This is helpful in case you want to create your own analysis tools or visualizations._
```
Comment (captainbrosset) [inline microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md:12]: Maybe try to make this a bit more compelling. Why would someone want to create their own visualizations of a heapsnapshot? Maybe give an example if possible.

Or rephrase to something like

"Investigating memory usage in web applications can be hard. The DevTools Memory tool helps with it, but you may sometimes need to focus on specific parts of the memory data that the Memory tool doesn't show. DevTools makes it possible to extract the entire memory data as a `*.heapsnapshot` file which you can then analyze and visualize. This article describes the structure and content of heapsnapshot files so you can build your own tools."

**Review by captainbrosset — CHANGES_REQUESTED** [review]

File: microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+All snapshots are dynamically generated based on the content of the web site or application, so the contents of the file will change. The code that generates this file lives here: [HeapSnapshotGenerator](https://chromium.googlesource.com/external/v8/+/master/src/heap-snapshot-generator.h)
+
+As the platform evolves there are bound to be changes in the format of this file, we always try to keep our documentation up to date but if you ever encounter a discrepancy this is your best bet to start your investigation. 
+
+This file has two main parts: **metadata** and **payload**. The metadata contains all the information you need to parse the different arrays of information in the payload and re-create the graph.
+
+### Base hierarchy of the file
+```JSON
+{
+    # This section contains all the metadata
```
[the remark this answers is not in the document]
Comment (joselea) [reply microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md:43]: ok

File: microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md (hunk 4 above)
Comment (joselea) [reply microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md:200]: woops, that came from an internal email from Seth, we shouldnt use that language in public

File: microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md (hunk 1 above)
Comment (joselea) [reply microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md:16]: a link to the v8 doc could be better

File: microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+meta | Contains several fields with information about the shape and size of every object contained in the payload  
+node_count | An integer representing the total of nodes 
+edge_count | An integer representing the total of edges. 
+trace_function_count | An integer representing the total of trace functions 
+nodes | An array with all the information requires to re-create the nodes | To parse this you need `metadata.node_types` and `metadata.node_fields`
+edges | An array with all the information required to re-create the edges | To parse this you need `metadata.edge_types` and `metadata.edge_fields`
+trace_function_infos | - |
+trace_tree | - |
+samples | - |
+locations| - |
```
[the remark this answers is not in the document]
Comment (joselea) [reply microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md:69]: add it later

File: microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md (hunk 5 above)
Comment (joselea) [reply microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md:202]: started a thread to find out where  is the best place to host it publicly.

File: microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (captainbrosset) [reply microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md:69]: Sounds good. Please ping me when you think you are done with this (and any other changes you're planning on making). I think on the content-side, we're good, so once you're done, I can take it from there and do an edit pass.

[the remark this answers is not in the document]
Comment (joselea) [reply microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md:69]: I meant later as in a second PR in the future

**Comment by captainbrosset** [comment]
Next steps:

1. Merge this into a new branch local to this repository (not main).
2. Do an edit pass on the new branch.
3. Do the final review and merge to main.
